README.md
[![example workflow](https://github.com/ephmoc/addonsly/actions/workflows/dart.yml/badge.svg)]() [![pub package][pub_badge]][pub_link] ![License: MIT](https://img.shields.io/github/license/ephmoc/addonsly) [![style: very good analysis](https://img.shields.io/badge/style-very_good_analysis-B22C89.svg)](https://pub.dev/packages/very_good_analysis) [![codecov](https://codecov.io/gh/ephmoc/addonsly/branch/main/graph/badge.svg?token=NYHZEAZE2C)](https://codecov.io/gh/ephmoc/addonsly)

# Addonsly

A package of useful Dart and Flutter addons

## Features

The package includes the following features:

- widgets
- class extensions (for `Object`, `String`, `List` etc.)
- UI constants

### Widgets

- Horizontal and vertical spacers (HSpacer and VSpacer) for UI content layout that extends
  `SizedBox` widget,
- `RowWithColumns` widget that contains a given number of widgets separated by a spacer

### Extensions

`List<Widget>` extensions like `joinWithSeparator()` method that allows to add spacers or other
widgets between widgets from a list.

`String` extensions like:

- `nullify` getter which return `null` if the string is empty
- `capitalized` or `firstCapitalLetter` getters for some basic string operations.

`Object` extensions like `unwrappedString` getter that returns a `String` representation of
an `Object` or `-` if it is `null` or an empty `String`

### Constants

`Paddings` set of `double` values useful for UI dimensions fine tuning.


[pub_badge]: https://img.shields.io/pub/v/addonsly.svg?label=pub

[pub_link]: https://pub.dev/packages/addonsly



CHANGELOG.md
## 0.0.5

* Upgraded dependencies

## 0.0.4

* Added `RowWithColumns` widget that contains a given number of widgets separated by a spacer

## 0.0.3

* Minor improvements

## 0.0.2

* updated Readme.md

## 0.0.1

* Initial version that brings widgets list and string extensions


test/src/src_test.dart
import 'extensions/extensions_test.dart' as extensions;
import 'widgets/row_with_columns_test.dart' as row_with_columns;
import 'widgets/spacers_test.dart' as spacers;

void main() {
  extensions.main();
  spacers.main();
  row_with_columns.main();
}


test/addonsly_test.dart
import 'src/src_test.dart' as tests;

void main() => tests.main();


test/src/widgets/spacers_test.dart
import 'package:addonsly/addonsly.dart';
import 'package:flutter/src/widgets/basic.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('VSpacers test', () {
    final vSpacers = {
      2.0: VSpacer.xxSmall(),
      4.0: VSpacer.xSmall(),
      6.0: VSpacer.small(),
      8.0: VSpacer.normal(),
      10.0: VSpacer.semiNormal(),
      12.0: VSpacer.semiMedium(),
      16.0: VSpacer.medium(),
      20.0: VSpacer.mediumLarge(),
      24.0: VSpacer.mediumXLarge(),
      32.0: VSpacer.large(),
      36.0: VSpacer.xLarge(),
      48.0: VSpacer.xxLarge(),
      64.0: VSpacer.xxxLarge(),
    };
    for (final pair in vSpacers.entries) {
      testWidgets('VSpacer test: ${pair.value}', (tester) async {
        final widget = pair.value;
        await tester.pumpWidget(Center(child: widget));
        final finder = find.byWidget(widget);
        expect(finder, findsOneWidget);
        final baseSize = tester.getSize(finder);
        expect(baseSize.height, equals(pair.key));
        expect(baseSize.width, equals(0));
      });
    }
  });

  group('HSpacers test', () {
    final hSpacers = {
      2.0: HSpacer.xxSmall(),
      4.0: HSpacer.xSmall(),
      6.0: HSpacer.small(),
      8.0: HSpacer.normal(),
      10.0: HSpacer.semiNormal(),
      12.0: HSpacer.semiMedium(),
      16.0: HSpacer.medium(),
      20.0: HSpacer.mediumLarge(),
      24.0: HSpacer.mediumXLarge(),
      32.0: HSpacer.large(),
      36.0: HSpacer.xLarge(),
      48.0: HSpacer.xxLarge(),
      64.0: HSpacer.xxxLarge(),
    };
    for (final pair in hSpacers.entries) {
      testWidgets('HSpacer test: ${pair.value}', (tester) async {
        final widget = pair.value;
        await tester.pumpWidget(Center(child: widget));
        final finder = find.byWidget(widget);
        expect(finder, findsOneWidget);
        final baseSize = tester.getSize(finder);
        expect(baseSize.width, equals(pair.key));
        expect(baseSize.height, equals(0));
      });
    }
  });
}


lib/src/constants/paddings.dart
/// Defines the default [double] values to be used as padding sizes
abstract class Paddings {
  /// padding of [2.0]
  static const double xxSmall = 2;

  /// padding of [4.0]
  static const double xSmall = 4;

  /// padding of [6.0]
  static const double small = 6;

  /// padding of [8.0]
  static const double normal = 8;

  /// padding of [10.0]
  static const double semiNormal = 10;

  /// padding of [12.0]
  static const double semiMedium = 12;

  /// padding of [16.0]
  static const double medium = 16;

  /// padding of [20.0]
  static const double mediumLarge = 20;

  /// padding of [24.0]
  static const double mediumXLarge = 24;

  /// padding of [32.0]
  static const double large = 32;

  /// padding of [36.0]
  static const double xLarge = 36;

  /// padding of [48.0]
  static const double xxLarge = 48;

  /// padding of [64.0]
  static const double xxxLarge = 64;
}


lib/src/widgets/row_with_columns.dart
import 'package:addonsly/addonsly.dart';
import 'package:collection/collection.dart';
import 'package:flutter/widgets.dart';

/// A [Row] that contains a given number of widgets separated by a spacer.
class RowWithColumns extends StatelessWidget {
  /// A [Row] that contains a given number of widgets separated by a spacer.
  /// `children` represents the list o widgets, `columnCount` is a number of
  /// columns in the row, `spacer` is an optional [Widget] the will be placed
  /// between each widget from the list (default: `HSpacer.semiNormal()`).
  const RowWithColumns({
    required this.children,
    required this.columnCount,
    super.key,
    this.spacer,
    this.crossAxisAlignment,
    this.mainAxisAlignment,
  }) : assert(columnCount > 0, 'columnCount must be greater then 0');

  /// A [Row] that contains two widgets separated by a spacer.
  /// `leading` is a first widget, `trailing` is a second one,
  /// `spacer` is an optional [Widget] the will be placed
  /// between each widget from the list (default: `HSpacer.semiNormal()`).
  RowWithColumns.double({
    super.key,
    Widget? leading,
    Widget? tailing,
    this.spacer,
    this.crossAxisAlignment,
    this.mainAxisAlignment,
  })  : children = [leading, tailing],
        columnCount = 2;

  /// A [Row] that contains three widgets separated by a spacer.
  /// `leading` is a first widget, `center` is a second one,
  /// `trailing` is a third, `spacer` is an optional [Widget] the will be
  /// placed between each widget from the list (default:
  /// `HSpacer.semiNormal()`).
  RowWithColumns.triple({
    super.key,
    Widget? leading,
    Widget? center,
    Widget? tailing,
    this.spacer,
    this.crossAxisAlignment,
    this.mainAxisAlignment,
  })  : children = [leading, center, tailing],
        columnCount = 3;

  /// Returns a [List] of [RowWithColumns] widgets (that could be used, e.g, in
  /// a [Column]).
  /// `columnCount` defines number of widgets in each row,
  /// `spacer` is an optional [Widget] the will be placed
  /// between each widget from the list (default: `HSpacer.semiNormal()`).
  static List<RowWithColumns> multipleFromList({
    required List<Widget?> children,
    required int columnCount,
    Widget? spacer,
    CrossAxisAlignment? crossAxisAlignment,
    MainAxisAlignment? mainAxisAlignment,
  }) {
    return children
        .slices(columnCount)
        .map(
          (slice) => RowWithColumns(
            key: ValueKey(slice),
            columnCount: columnCount,
            crossAxisAlignment: crossAxisAlignment,
            mainAxisAlignment: mainAxisAlignment,
            spacer: spacer,
            children: slice.length == columnCount
                ? slice
                : [
                    ...slice,
                    for (var i = slice.length; i < columnCount; i++) null
                  ],
          ),
        )
        .toList();
  }

  /// A list of widgets placed in the row.
  final List<Widget?> children;

  /// A number of widgets placed in the row.
  final int columnCount;

  /// A spacer placed between each widget in the row.
  final Widget? spacer;

  /// How the children should be placed along the cross axis in a flex layout.
  final CrossAxisAlignment? crossAxisAlignment;

  /// How the children should be placed along the main axis in a flex layout.
  final MainAxisAlignment? mainAxisAlignment;

  @override
  Widget build(BuildContext context) {
    return Row(
      mainAxisAlignment: mainAxisAlignment ?? MainAxisAlignment.start,
      crossAxisAlignment: crossAxisAlignment ?? CrossAxisAlignment.start,
      children: children
          .map((w) => Expanded(child: w ?? const SizedBox.shrink()))
          .toList()
          .joinWithSeparator(spacer ?? HSpacer.semiNormal()),
    );
  }
}


lib/src/extensions/string_extensions.dart
import 'package:collection/collection.dart';
import 'package:flutter/widgets.dart';

/// A collection of useful extensions on [String]
extension AddonslyString on String {
  /// A getter that converts the first letter of a [String] into uppercase
  /// and the remaining letters of the [String] into lowercase
  /// if the [String] is not empty.
  ///
  /// Example:
  /// ```dart
  /// 'john Doe'.capitalized; // John doe
  /// ```
  String get capitalized => isNotEmpty
      ? '${this[0].toUpperCase()}${substring(1).toLowerCase()}'
      : this;

  /// Returns a first letter of a [String] converted to uppercase
  /// or null if the [String] is empty
  String? get firstCapitalLetter => characters.firstOrNull?.toUpperCase();
}

/// A collection of useful extensions on [String?]
extension AddonslyNullableString on String? {
  /// Returns null if a [String] is empty
  String? get nullify {
    if (this?.isEmpty ?? true) {
      return null;
    } else {
      return this;
    }
  }
}


lib/src/extensions/list_widget_extensions.dart
import 'package:flutter/widgets.dart';

/// A collection of useful extensions on [List<Widget>]
extension AddonslyWidgetList on List<Widget> {
  /// Adds a specified separator [Widget] between Widgets in a list.
  /// This can be useful to add some spacer or other kind of divider between
  /// Widgets without the need of writing it multiple times and cluttering the
  /// code.
  ///
  /// Example:
  /// Instead of writing this:
  /// ```dart
  /// [
  ///  const Text('AAA'),
  ///  const Divider(),
  ///  const Text('BBB'),
  ///  const Divider(),
  ///  const Text('CCC'),
  /// ]
  /// ```
  ///
  /// you can write:
  ///
  /// ```dart
  /// [
  ///  const Text('AAA'),
  ///  const Text('BBB'),
  ///  const Text('CCC'),
  /// ].joinWithSeparator(const Divider());
  /// ```
  List<Widget> joinWithSeparator(Widget separator) {
    return length > 1
        ? (take(length - 1)
            .map((widget) => [widget, separator])
            .expand((widget) => widget)
            .toList()
          ..add(last))
        : this;
  }
}


lib/src/widgets/spacers.dart
import 'package:addonsly/src/constants/paddings.dart';
import 'package:flutter/widgets.dart';

/// [VSpacer] creates a vertical separation between Widgets
/// the same way you would use a SizedBox, but with predefined values
/// from [Paddings]
///
/// Example:
/// ```dart
///    VSpacer.xxSmall(),
///    VSpacer.xSmall(),
///    VSpacer.small(),
///    VSpacer.normal(),
///    VSpacer.semiNormal(),
///    VSpacer.semiMedium(),
///    VSpacer.medium(),
///    VSpacer.mediumLarge(),
///    VSpacer.mediumXLarge(),
///    VSpacer.large(),
///    VSpacer.xLarge(),
///    VSpacer.xxLarge(),
///    VSpacer.xxxLarge(),
/// ```
///
/// See also:
///
///  * [Paddings]
///  * [HSpacer]
class VSpacer extends SizedBox {
  /// [VSpacer] with custom paddings
  const VSpacer({required double super.height, super.key});

  /// [VSpacer] with [2.0] as padding
  factory VSpacer.xxSmall() => const VSpacer(height: Paddings.xxSmall);

  /// [VSpacer] with [4.0] as padding
  factory VSpacer.xSmall() => const VSpacer(height: Paddings.xSmall);

  /// [VSpacer] with [6.0] as padding
  factory VSpacer.small() => const VSpacer(height: Paddings.small);

  /// [VSpacer] with [8.0] as padding
  factory VSpacer.normal() => const VSpacer(height: Paddings.normal);

  /// [VSpacer] with [10.0] as padding
  factory VSpacer.semiNormal() => const VSpacer(height: Paddings.semiNormal);

  /// [VSpacer] with [12.0] as padding
  factory VSpacer.semiMedium() => const VSpacer(height: Paddings.semiMedium);

  /// [VSpacer] with [16.0] as padding
  factory VSpacer.medium() => const VSpacer(height: Paddings.medium);

  /// [VSpacer] with [20.0] as padding
  factory VSpacer.mediumLarge() => const VSpacer(height: Paddings.mediumLarge);

  /// [VSpacer] with [24.0] as padding
  factory VSpacer.mediumXLarge() =>
      const VSpacer(height: Paddings.mediumXLarge);

  /// [VSpacer] with [32.0] as padding
  factory VSpacer.large() => const VSpacer(height: Paddings.large);

  /// [VSpacer] with [36.0] as padding
  factory VSpacer.xLarge() => const VSpacer(height: Paddings.xLarge);

  /// [VSpacer] with [48.0] as padding
  factory VSpacer.xxLarge() => const VSpacer(height: Paddings.xxLarge);

  /// [VSpacer] with [64.0] as padding
  factory VSpacer.xxxLarge() => const VSpacer(height: Paddings.xxxLarge);
}

/// [HSpacer] creates a horizontal separation between Widgets
/// the same way you would use a SizedBox, but with predefined values
/// from [Paddings]
///
/// Example:
/// ```dart
///    HSpacer.small()
///    HSpacer.normal()
///    HSpacer.semi()
///    HSpacer.mediumSmall()
///    HSpacer.medium()
///    HSpacer.medium20()
///    HSpacer.mediumLarge()
///    HSpacer.large()
///    HSpacer.xLarge()
///    HSpacer.xxLarge()
/// ```
///
/// See also:
///
///  * [Paddings]
///  * [VSpacer]
class HSpacer extends SizedBox {
  /// [HSpacer] with custom paddings
  const HSpacer({required double super.width, super.key});

  /// [HSpacer] with [2.0] as padding
  factory HSpacer.xxSmall() => const HSpacer(width: Paddings.xxSmall);

  /// [HSpacer] with [4.0] as padding
  factory HSpacer.xSmall() => const HSpacer(width: Paddings.xSmall);

  /// [HSpacer] with [6.0] as padding
  factory HSpacer.small() => const HSpacer(width: Paddings.small);

  /// [HSpacer] with [8.0] as padding
  factory HSpacer.normal() => const HSpacer(width: Paddings.normal);

  /// [HSpacer] with [10.0] as padding
  factory HSpacer.semiNormal() => const HSpacer(width: Paddings.semiNormal);

  /// [HSpacer] with [12.0] as padding
  factory HSpacer.semiMedium() => const HSpacer(width: Paddings.semiMedium);

  /// [HSpacer] with [16.0] as padding
  factory HSpacer.medium() => const HSpacer(width: Paddings.medium);

  /// [HSpacer] with [20.0] as padding
  factory HSpacer.mediumLarge() => const HSpacer(width: Paddings.mediumLarge);

  /// [HSpacer] with [24.0] as padding
  factory HSpacer.mediumXLarge() => const HSpacer(width: Paddings.mediumXLarge);

  /// [HSpacer] with [32.0] as padding
  factory HSpacer.large() => const HSpacer(width: Paddings.large);

  /// [HSpacer] with [36.0] as padding
  factory HSpacer.xLarge() => const HSpacer(width: Paddings.xLarge);

  /// [HSpacer] with [48.0] as padding
  factory HSpacer.xxLarge() => const HSpacer(width: Paddings.xxLarge);

  /// [HSpacer] with [64.0] as padding
  factory HSpacer.xxxLarge() => const HSpacer(width: Paddings.xxxLarge);
}
